### ISSUE TITLE
[css-overflow] [css2?] Baseline of an inline-block with overflow:clip.

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#6212]
Blink and Gecko differ on what baseline should an element with `display: inline-block` and `overflow: clip`.

Example ([live](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/9184)):

```html
<!doctype html>

Clip:
<br>
<div style="border:1px solid red; display:inline-block">
  <div style="overflow:clip; width:10px; height:10px; background:blue; display:inline-block;">Hallo Welt</div>
</div>

<br>
Visible:
<br>
<div style="border:1px solid red; display:inline-block">
  <div style="width:10px; height:10px; background:blue; display:inline-block;">Hallo Welt</div>
</div>
<br>
Hidden:
<br>
<div style="border:1px solid red; display:inline-block">
  <div style="overflow:hidden; width:10px; height:10px; background:blue; display:inline-block;">Hallo Welt</div>
</div>
```

@bfgeek pointed out that CSS2.1 has:

>  The baseline of an 'inline-block' is the baseline of its last line box in the normal flow, unless it has either no in-flow line boxes or if its 'overflow' property has a computed value other than 'visible', in which case the baseline is the bottom margin edge.

However CSS2 was written before `overflow: clip` was a thing. I intuitively expect `overflow: clip` to behave like `overflow: visible` here (because it's not an scrollable box), but I could see the argument the other way around. Thoughts?

**Comment by emilio** [comment]
An interesting question, if we decide that `clip` and `visible` should behave differently, is what happens if only one of the axes is clipped.

**Comment by bfgeek** [comment]
I somewhat expected blink's behaviour (and have a mild preference for it) as I sort of view `clip` as a better `hidden`, but either way is fine.

**Comment by Loirooriol** [comment]
> what happens if only one of the axes is clipped

Seems a good argument for treating `clip` like `visible`.

**Comment by tabatkins** [comment]
Yeah, since only one axis can be clipped, this really should continue to be treated like `visible`, not like `hidden`. Diverging from that principle will only cause us confusion in the future, I think.

And yes, older text saying "overflow other than visible" just needs to be fixed up, it should always *mean* "scroll container" in modern parlance.

**Comment by MatsPalmgren** [comment]
I agree,  `overflow-inline : clip` should not affect the baseline, so that disqualifies the "behave as `hidden`" option as described above. However, I don't see a reason why we can't synthesize a baseline for `overflow-block : clip` (which would behave roughly as `hidden` in that axis).

It's also worth considering what happens when the baseline is well within the size, e.g. emilio's test [with `height:100px` instead](http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=9191). I'd slightly prefer the `visible` behavior over a synthesized baseline in this case, but I don't feel strongly about it.

There's also a third option: "clamp the baseline", which would calculate the baseline normally (as for `visible`) but clamp it to the clip area.  That would make emilio's test behave (roughly*) as `hidden` but the `height:100px` case would still behave as `visible`. (I dunno how complicated that would be to implement though, so I'm not sure if it's worth it.)

(*) the clip area is typically the padding box, whereas the `hidden` baseline is calculated from the border box

So I see three viable options:
1. the baseline is calculated the same as for `visible` (what Firefox currently implements)
2. always synthesize a baseline for `overflow-block : clip`
3. as 1 but clamp it to the clip area

**Comment by bernhardf-ro** [comment]
Thanks @MatsPalmgren for the list of options. I'll reference it for brevity.
In PDFreactor we have implemented a variation of 3: For `clip` we clamp the `visible` baseline to the synthesized baseline. So you always have either 1 or 2 (`visible` or `hidden` behavior) in the end.
We considered the clip area as clamp value, but concluded that the `hidden` behavior would be more expected than adding a new area affecting alignment.
Maybe you can add this option for discussion as 4 or 3b.

**Comment by bernhardf-ro** [comment]
We can change the behavior of PDFreactor accordingly.
Clamping was not required, just what we came up with internally.

**Comment by khushalsagar** [comment]
Could the spec be updated based on this resolution?

**Comment by dbaron** [comment]
It's worth noting that the relevant level 3 spec here is [css-align-3](https://drafts.csswg.org/css-align-3/#baseline-export), which both (a) has this issue fixed already and (b) makes a separate incompatible change regarding clamping the baseline to the border edge rather than always using the margin edge:

> For the purposes of finding the baselines of a box, it and all its in-flow descendants with a scrolling mechanism (see the [overflow](https://drafts.csswg.org/css-overflow-3/#propdef-overflow) property) must be considered as if scrolled to their initial scroll position. Furthermore, if, in the case of a [scroll container](https://drafts.csswg.org/css-overflow-3/#scroll-container) box, the resulting position of a first/last baseline is past a box’s end/start border edge, its position is clamped to that border edge.
